src/Storages/MergeTree/PartitionPruner.h

```diff
@@ -0,0 +1,53 @@
+#pragma once
+
+#include <unordered_map>
+
+#include <Storages/KeyDescription.h>
+#include <Storages/MergeTree/IMergeTreeDataPart.h>
+#include <Storages/MergeTree/KeyCondition.h>
+
+namespace DB
+{
+
+class PartitionPruner
+{
+private:
+    std::unordered_map<String, bool> partition_filter_map;
+    const KeyDescription & partition_key;
+    KeyCondition partition_condition;
+    bool useless;
+    using DataPart = IMergeTreeDataPart;
+    using DataPartPtr = std::shared_ptr<const DataPart>;
+
+public:
+    PartitionPruner(const KeyDescription & partition_key_, const SelectQueryInfo & query_info, const Context & context)
+        : partition_key(partition_key_)
+        , partition_condition(
+              query_info, context, partition_key.column_names, partition_key.expression, true /* single_point */, true /* strict */)
+        , useless(partition_condition.alwaysUnknownOrTrue())
+    {
+    }
+
+    bool canBePruned(DataPartPtr part)
```

Comment:
Better pass by const reference (because it's not moved, it's not modified and the operation with atomic counter is non-free).